### ISSUE TITLE
Is there a tool that can separate EMA and non-EMA weights and switch VAE and Convert safetensors to ckpt 

### ISSUE THREAD

**Issue by camenduru** [huggingface/safetensors#149]
Good news, people started sharing only safetensors 🎉, but do we have tools to modify safetensors like ckpt?

we need  
- separate EMA and non-EMA weights tool 
- switch VAE inside  safetensors  tool 
- convert safetensors to ckpt tool

Maybe these tools are already available, I don't know where. 🤔

**Comment by pedx78** [comment]
3 - convert safetensors to ckpt tool

https://github.com/diStyApps/Safe-and-Stable-Ckpt2Safetensors-Conversion-Tool-GUI

https://github.com/huggingface/safetensors/issues/120

**Comment by camenduru** [comment]
thanks @pedx78 ♥

**Comment by Narsil** [comment]
Not sure I understand everything here.

I mean what would exactly be missing as tooling ?

**Comment by camenduru** [comment]
Hi @Narsil 👋  how can I separate ema weights from the safetensors model

**Comment by Narsil** [comment]
Do you have a checkpoints I could look at that contain both ? And maybe the piece of could that is EMA specific.

I try to keep up with SD but unfortunately can't especially on the workflows part so I know EMA are a thing, embeddings are a thing, but not having used them I'm not sure what are we talking about.

If you're not sure, it's ok just do tell and I'll investigate, but if you can help me figure it out faster, I'll take it :)

**Comment by camenduru** [comment]
yep like this one https://huggingface.co/deadman44/SD_Anime_Merged_Models/blob/main/El%20Diablo.safetensors

**Comment by Narsil** [comment]
I see the same model twice in there:

`model.xx.yy....`
and 
`model_ema.xx.yy....`

Then isn't it just a matter of 

```python
from safetensors.torch import load_file, save_file

tensors = load_file("eldiablo.safetensors")
filtered_only_ema = {k: v for k, v in tensors.items() if not k.startswith("model.")}
save_file(filtered_only_ema, "eldiablo-emaonly.safetensors'?
```

**Comment by camenduru** [comment]
thanks @Narsil ❤ I will try it

**Comment by camenduru** [comment]
Hi @Narsil 👋  what should I do when I hit the `HeaderTooLarge` 😐 also why 100_000_000  

`const MAX_HEADER_SIZE: usize = 100_000_000;`

`!wget "https://civitai.com/api/download/models/4298?type=Model&format=SafeTensor"`

```py
[/usr/local/lib/python3.8/dist-packages/safetensors/torch.py](https://localhost:8080/#) in load_file(filename, device)
     96     """
     97     result = {}
---> 98     with safe_open(filename, framework="pt", device=device) as f:
     99         for k in f.keys():
    100             result[k] = f.get_tensor(k)

Exception: Error while deserializing header: HeaderTooLarge
```

**Comment by Narsil** [comment]
100Mo for the JSON part is huge and unlikely to be hit on a 7Go file (except if you have 1 elements tensors everywhere for instance).

The rationale for having a size limit is to prevent DOS attacks with the file (forcing you to allocate much more than necessary by sending a humongously large JSON being hard to parse). This can be relaxed later if real needs really arise or maybe made controllable by some kind of environment variable later if needed.  

However, it seems like a bitflip in the initial part of the file (which contains the size) since I'm able to open the file without any issue:

I'm getting

```
sha256sum test.safetensors
6a21b428a3fb7286f024f958c761ea1a36a5061c3d3c1eb6a815c88af0e97cb0  test.safetensors
```
The header size is 249kb (so looong way to go for 100Mo :) ).

Is this the correct file ?

**Comment by camenduru** [comment]
oops wrong file 😋

**Comment by camenduru** [comment]
Maybe the file is corrupted or something, I can't regenerate the error 😋 it is working. I will add here: https://huggingface.co/spaces/camenduru/converter. thanks @Narsil ❤

**Comment by camenduru** [comment]
and the vae thing I am switching the vae like this https://github.com/camenduru/notebooks/blob/main/camenduru's_replace_vae.ipynb
is it possible for safetensors

**Comment by camenduru** [comment]
I think first we need to convert this type https://huggingface.co/andite/anything-v4.0/blob/main/anything-v4.0.vae.pt to safetensors

**Comment by camenduru** [comment]
and this type https://huggingface.co/datasets/Nerfgun3/bad_prompt/blob/main/bad_prompt_version2.pt

**Comment by Narsil** [comment]
Sorry, pretty hard to follow, do you mind opening a new issue with the new problem ? It doesn't seem really related to this correct? 

Most of the stuff in your notebook is just weights/weight name manipulation, which you can probably do just the same with `safetensors`.

**Comment by camenduru** [comment]
ok I will open later thanks ❤
